### PR TITLE
 Implemented the process of downloading the assembly sequences fasta file

### PR DESCRIPTION
**Pull request by waterflow80**
Implemented the process of downloading the assembly sequences fasta file

Created and tested the classes for the following:

    1. Downloand the assembly sequences' fasta file from NCBI database
    2. Unzip the _fna.gz file
    3. Parse the fasta file, retrieve the sequences and hash them with md5 algorithm
    4. Save the sequences into the AssemblySequencesEntity class

### CODE REVIEW

File: src/main/java/uk/ac/ebi/eva/evaseqcol/datasource/NCBIAssemblySequencesDataSource.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.security.NoSuchAlgorithmException;
+import java.util.Optional;
+
+@Repository("NCBISequencesDataSource")
+public class NCBIAssemblySequencesDataSource implements AssemblySequencesDataSource{
```
Comment (sundarvenkata-EBI) [inline src/main/java/uk/ac/ebi/eva/evaseqcol/datasource/NCBIAssemblySequencesDataSource.java:29]: Suggest renaming NCBIAssemblySequencesDataSource to NCBIAssemblySequenceDataSource

File: src/main/java/uk/ac/ebi/eva/evaseqcol/dus/AssemblySequencesReader.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
+package uk.ac.ebi.eva.evaseqcol.dus;
+
+import uk.ac.ebi.eva.evaseqcol.entities.AssemblySequencesEntity;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.security.NoSuchAlgorithmException;
+
+public abstract class AssemblySequencesReader {
```
Comment (sundarvenkata-EBI) [inline src/main/java/uk/ac/ebi/eva/evaseqcol/dus/AssemblySequencesReader.java:10]: Suggest renaming to AssemblySequenceReader

File: src/main/java/uk/ac/ebi/eva/evaseqcol/datasource/NCBIAssemblySequencesDataSource.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+            return Optional.empty();
+        }
+
+        logger.info("NCBI directory for assembly genomic.fna download: " + directory.get());
+        FTPFile ftpFile = ncbiBrowser.getAssemblySequencesFastaFile(directory.get());
+        String ftpFilePath = directory.get() + ftpFile.getName();
+        Path downloadFilePath = Paths.get(asmFileDownloadDir, ftpFile.getName());
+        boolean success = ncbiBrowser.downloadFTPFile(ftpFilePath, downloadFilePath, ftpFile.getSize());
+        if (success) {
+            logger.info("NCBI assembly genomic.fna downloaded successfully (" + ftpFile.getName() + ")");
```
Comment (sundarvenkata-EBI) [inline src/main/java/uk/ac/ebi/eva/evaseqcol/datasource/NCBIAssemblySequencesDataSource.java:100]: `logger.info("NCBI assembly FASTA downloaded successfully (" + ftpFile.getName() + ")");`

File: src/main/java/uk/ac/ebi/eva/evaseqcol/datasource/NCBIAssemblySequencesDataSource.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+            return Optional.of(downloadFilePath);
+        } else {
+            logger.error("NCBI assembly genomic.fna could not be downloaded successfully(" + ftpFile.getName() + ")");
```
Comment (sundarvenkata-EBI) [inline src/main/java/uk/ac/ebi/eva/evaseqcol/datasource/NCBIAssemblySequencesDataSource.java:103]: `logger.error("NCBI assembly genomic.fna could not be downloaded successfully(" + ftpFile.getName() + ")");`

File: src/main/java/uk/ac/ebi/eva/evaseqcol/dus/AssemblySequencesReader.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    protected final String accession;
+
+    protected AssemblySequencesEntity assemblySequencesEntity;
+
+
+    protected boolean fileParsed = false;
+
+
+    public AssemblySequencesReader(InputStreamReader inputStreamReader, String accession){
+        this.reader = new BufferedReader(inputStreamReader);
```
Comment (sundarvenkata-EBI) [inline src/main/java/uk/ac/ebi/eva/evaseqcol/dus/AssemblySequencesReader.java:22]: **Optional:** This reader object seems to be closed manually in the inheritees.Check if it is possible to provide a top-level method here that will call parseFile in the inheritees (a method prone to exceptions) and close the reader after parseFile exits.

File: src/main/java/uk/ac/ebi/eva/evaseqcol/dus/NCBIAssemblySequencesReader.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+                    line = reader.readLine();
+                }
+                String md5checksum = md5Digest.hash(sequenceValue.toString());
+                sequence.setSequenceMD5(md5checksum);
+                sequences.add(sequence);
+            }
+        }
+        assemblySequencesEntity.setSequences(sequences);
+        fileParsed = true;
+        reader.close();
```
Comment (sundarvenkata-EBI) [inline src/main/java/uk/ac/ebi/eva/evaseqcol/dus/NCBIAssemblySequencesReader.java:50]: See comment [here](https://github.com/EBIvariation/eva-seqcol/pull/2/commits/2db1dd676dfaee92f5701034eaf37f5604277734#r1246280713)

File: src/main/java/uk/ac/ebi/eva/evaseqcol/dus/NCBIBrowser.java [hunk 7]
```diff
@@ -131,4 +131,14 @@ public FTPFile getNCBIAssemblyReportFile(String directoryPath) throws IOExceptio
 
         return assemblyReport.orElseThrow(() -> new AssemblyNotFoundException("Assembly Report File not present in given directory: " + directoryPath));
     }
+
+    /**
+     * Return a pointer to the assembly sequences' fasta file that will be downloaded*/
+    public FTPFile getAssemblySequencesFastaFile(String directoryPath) throws IOException {
+        Stream<FTPFile> ftpFileStream = Arrays.stream(super.listFiles(directoryPath));
+        Stream<FTPFile> assemblyReportFilteredStream = ftpFileStream.filter(f -> f.getName().contains("genomic.fna.gz") && !f.getName().contains("from"));
```
Comment (sundarvenkata-EBI) [inline src/main/java/uk/ac/ebi/eva/evaseqcol/dus/NCBIBrowser.java:139]: `Stream<FTPFile> assemblyReportFilteredStream = ftpFileStream.filter(f -> f.getName().toLowerCase().contains("genomic.fna.gz") && !f.getName().toLowerCase().contains("from"));`

Comment (sundarvenkata-EBI) [reply src/main/java/uk/ac/ebi/eva/evaseqcol/dus/NCBIBrowser.java:139]: btw, I am wondering why the "from" filter is needed?

**Review by sundarvenkata-EBI — APPROVED** [review]
Looks good! Optional improvements are marked as such.

File: src/main/java/uk/ac/ebi/eva/evaseqcol/datasource/NCBIAssemblySequencesDataSource.java (hunk 1 above)
Comment (waterflow80) [reply src/main/java/uk/ac/ebi/eva/evaseqcol/datasource/NCBIAssemblySequencesDataSource.java:29]: Okay I'll do it.

File: src/main/java/uk/ac/ebi/eva/evaseqcol/dus/NCBIBrowser.java (hunk 7 above)
Comment (waterflow80) [reply src/main/java/uk/ac/ebi/eva/evaseqcol/dus/NCBIBrowser.java:139]: @sundarvenkata-EBI @tcezard 
Some assemblies have **multiple** Fasta files in their home path that ends with _genomic_fna.gz.
Eg [here](https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/405/GCF_000001405.40_GRCh38.p14/)

ambiguous-fasta

So I assumed that the one we're concerned about is the one that ends only with _**genomic_fna.gz (no _from_..)**, which also has the largest file size, based on other assemblies that has only one Fasta file and its name ends with **_genomic_fna.gz**, like this [one](https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/405/GCF_000001405.29_GRCh38.p3/).

It was a self observation, but I think I should have raised an issue about it just to make sure.

Comment (tcezard) [reply src/main/java/uk/ac/ebi/eva/evaseqcol/dus/NCBIBrowser.java:139]: I think we'll see that in GCF assembly much more than in GCA but it makes sense to keep the filter for now.
Can you just add a comment with a link to the example assembly for the next person wondering.

Comment (waterflow80) [reply src/main/java/uk/ac/ebi/eva/evaseqcol/dus/NCBIBrowser.java:139]: @tcezard 
Okay I will.
In the method's doc comment right ?

**Review by tcezard — APPROVED** [review]

**Review by tcezard — APPROVED** [review]

**Review by sundarvenkata-EBI — APPROVED** [review]

**Pull request merged**